File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_simple.py
from collections import deque

# Ham nhap do thi tu ban phim
def input_graph():
    n = int(input("Nhap so dinh: "))  # So dinh
    m = int(input("Nhap so canh: "))  # So canh
    adj = [[] for _ in range(n)]      # Danh sach ke
    for _ in range(m):
        u, v = map(int, input("Nhap canh u v: ").split())
        adj[u].append(v)
        adj[v].append(u)              # Do thi vo huong
    return n, adj

# Ham BFS
def bfs(start, adj, n):
    visited = [False] * n            # Mang danh dau da tham chua
    queue = deque([start])           # Hang doi
    visited[start] = True
    print("Thu tu duyet:", end=' ')
    while queue:
        u = queue.popleft()          # Lay dinh tu hang doi
        print(u, end=' ')            # In ra dinh dang duyet
        for v in adj[u]:
            if not visited[v]:       # Neu chua tham thi dua vao hang doi
                visited[v] = True
                queue.append(v)

# Chuong trinh chinh
if __name__ == "__main__":
    n, adj = input_graph()
    s = int(input("Nhap dinh bat dau: "))
    bfs(s, adj, n)

File: combinatorics/resource/ST/project_5/bai_toan_8_9_10/bfs_simple.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Ham nhap do thi
void input_graph(int& n, vector<vector<int>>& adj) {
    int m;
    cout << "Nhap so dinh: ";
    cin >> n;
    cout << "Nhap so canh: ";
    cin >> m;
    adj.resize(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cout << "Nhap canh u v: ";
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u); // Vo huong
    }
}

// Ham BFS
void bfs(int start, vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    cout << "Thu tu duyet: ";
    while (!q.empty()) {
        int u = q.front(); q.pop();
        cout << u << " ";
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

int main() {
    int n;
    vector<vector<int>> adj;
    input_graph(n, adj);
    int s;
    cout << "Nhap dinh bat dau: ";
    cin >> s;
    bfs(s, adj, n);
    return 0;
}